core/sessions/ldapauth/ldap.go

```diff
@@ -0,0 +1,865 @@
+		return "", fmt.Errorf("error creating local LDAP session: %w", err)
+	}
+
+	l.auditLogger.Audit(audit.AuthLoginSuccessNo2FA, map[string]interface{}{"email": sr.Email})
+
+	return session.ID, nil
+}
+
+// ClearNonCurrentSessions removes all ldap_sessions but the id passed in.
+func (l *ldapAuthenticator) ClearNonCurrentSessions(sessionID string) error {
+	_, err := l.q.Exec("DELETE FROM ldap_sessions where id != $1", sessionID)
+	return err
+}
+
+// CreateUser is not supported for read only LDAP
+func (l *ldapAuthenticator) CreateUser(user *sessions.User) error {
+	return sessions.ErrNotSupported
+}
+
+// UpdateRole is not supported for read only LDAP
+func (l *ldapAuthenticator) UpdateRole(email, newRole string) (sessions.User, error) {
+	return sessions.User{}, sessions.ErrNotSupported
+}
+
+// SetPassword for remote users is not supported via the read only LDAP implementation, however change password
+// in the context of updating a local admin user's password is required
+func (l *ldapAuthenticator) SetPassword(user *sessions.User, newPassword string) error {
+	// Ensure specified user is part of the local admins user table
+	var localAdminUser sessions.User
+	if err := l.q.Transaction(func(tx pg.Queryer) error {
+		sql := "SELECT * FROM users WHERE lower(email) = lower($1)"
+		return tx.Get(&localAdminUser, sql, user.Email)
+	}); err != nil {
+		l.lggr.Infof("Can not change password, local user with email not found in users table: %s, err: %v", user.Email, err)
+		return sessions.ErrNotSupported
+	}
+
+	// User is local admin, save new password
+	hashedPassword, err := utils.HashPassword(newPassword)
+	if err != nil {
+		return err
+	}
+	if err := l.q.Transaction(func(tx pg.Queryer) error {
+		sql := "UPDATE users SET hashed_password = $1, updated_at = now() WHERE email = $2 RETURNING *"
+		return tx.Get(user, sql, hashedPassword, user.Email)
+	}); err != nil {
+		l.lggr.Errorf("unable to set password for user: %s, err: %v", user.Email, err)
+		return errors.New("unable to save password")
+	}
+	return nil
+}
+
+// TestPassword tests if an LDAP login bind can be performed with provided credentials, returns nil if success
+func (l *ldapAuthenticator) TestPassword(email string, password string) error {
+	// Establish ephemeral connection
+	conn, err := l.ldapClient.CreateEphemeralClient()
+	if err != nil {
+		return errors.New("unable to establish connection to LDAP server with provided URL and credentials")
+	}
+	defer conn.Close()
+
+	// Attempt to LDAP Bind with user provided credentials
+	escapedEmail := ldap.EscapeFilter(strings.ToLower(email))
+	searchBaseDN := fmt.Sprintf("%s=%s,%s,%s", l.config.BaseUserAttr(), escapedEmail, l.config.UsersDN(), l.config.BaseDN())
+	err = conn.Bind(searchBaseDN, password)
+	if err != nil {
+		l.lggr.Infof("Error binding user authentication request in TestPassword call LDAP Bind: %v", err)
+	} else {
+		// LDAP Bind/login was successful, return success case
+		return nil
+	}
```

Comment:
Maybe this way reads a little easier?
```suggestion
	if err == nil {
		// LDAP Bind/login was successful, return success case
		return nil
	}
	l.lggr.Infof("Error binding user authentication request in TestPassword call LDAP Bind: %v", err)
```

Reply:
Yup, flipped logic